go/vt/discovery/healthcheck.go

```diff
@@ -412,44 +395,74 @@ func (hc *HealthCheckImpl) ServeHTTP(w http.ResponseWriter, _ *http.Request) {
 // servingConnStats returns the number of serving tablets per keyspace/shard/tablet type.
 func (hc *HealthCheckImpl) servingConnStats() map[string]int64 {
 	res := make(map[string]int64)
-	hc.mu.RLock()
-	defer hc.mu.RUnlock()
-	for _, hcc := range hc.addrToConns {
-		hcc.mu.RLock()
-		if !hcc.tabletStats.Up || !hcc.tabletStats.Serving || hcc.tabletStats.LastError != nil {
-			hcc.mu.RUnlock()
+	hc.mu.Lock()
+	defer hc.mu.Unlock()
+	for _, th := range hc.addrToHealth {
+		if !th.tabletStats.Up || !th.tabletStats.Serving || th.tabletStats.LastError != nil {
 			continue
 		}
-		key := fmt.Sprintf("%s.%s.%s", hcc.tabletStats.Target.Keyspace, hcc.tabletStats.Target.Shard, topoproto.TabletTypeLString(hcc.tabletStats.Target.TabletType))
-		hcc.mu.RUnlock()
+		key := fmt.Sprintf("%s.%s.%s", th.tabletStats.Target.Keyspace, th.tabletStats.Target.Shard, topoproto.TabletTypeLString(th.tabletStats.Target.TabletType))
 		res[key]++
 	}
 	return res
 }
 
+// updateHealth updates the tabletHealth record and transmits the tablet stats
+// to the listener.
+func (hc *HealthCheckImpl) updateHealth(ts *TabletStats, conn queryservice.QueryService) {
+	// Unconditionally send the received update at the end.
+	defer func() {
+		if hc.listener != nil {
+			hc.listener.StatsUpdate(ts)
+		}
+	}()
+
+	hc.mu.Lock()
+	th, ok := hc.addrToHealth[ts.Key]
+	if !ok {
+		// This can happen on delete because the entry is removed first.
+		hc.mu.Unlock()
+		return
+	}
+	oldts := th.tabletStats
+	th.tabletStats = *ts
```

Comment:
Instead of calling `copy()` on the way in, it seems more defensive to copy it here.